torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -406,30 +406,51 @@ def sync(self) -> Optional[bool]:
         return has_set
 
     def _sanitize(self) -> None:
+        state = self._state
+
         expire_time = datetime.utcnow() - (
             self._settings.keep_alive_interval * self._settings.keep_alive_max_attempt
         )
 
         # Filter out the dead nodes.
         self._dead_nodes = [
             node
-            for node, last_heartbeat in self._state.last_heartbeats.items()
+            for node, last_heartbeat in state.last_heartbeats.items()
             if last_heartbeat < expire_time
         ]
 
+        participant_removed = False
+
         for dead_node in self._dead_nodes:
-            del self._state.last_heartbeats[dead_node]
+            del state.last_heartbeats[dead_node]
 
             try:
-                del self._state.participants[dead_node]
+                del state.participants[dead_node]
+
+                participant_removed = True
             except KeyError:
                 pass
 
             try:
-                self._state.wait_list.remove(dead_node)
+                state.wait_list.remove(dead_node)
             except KeyError:
                 pass
 
+        if not participant_removed:
+            return
+
+        # If a participant is removed from the rendezvous, we have to make sure
+        # that the state of the rendezvous reflects it.
+        if state.complete:
+            # If we do not have any participants left, move to the next round.
+            if not state.participants:
+                state.complete = False
+
+                state.round += 1
+        else:
+            if len(state.participants) < self._settings.min_nodes:
+                state.deadline = None
```

Comment:
If we are here, it means the rendezvous is still open and has not reached the maximum number of nodes yet. So we are in the "lastcall" phase. However in case the number of ready nodes drop below the minimum, we should cancel the lastcall; otherwise we will potentially dispatch the rendezvous with a suboptimal number of nodes.